presto-jmx/src/main/java/com/facebook/presto/connector/jmx/JmxHistoryDumper.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.connector.jmx;
+
+import com.facebook.presto.spi.SchemaTableName;
+import com.google.inject.Inject;
+import io.airlift.log.Logger;
+
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+public class JmxHistoryDumper
+        implements Runnable
+{
+    private static final Logger log = Logger.get(JmxHistoryDumper.class);
+
+    private final JmxHistoricalData jmxHistoricalData;
+    private final JmxMetadata jmxMetadata;
+    private final JmxRecordSetProvider jmxRecordSetProvider;
+    private final ScheduledExecutorService executor = newSingleThreadScheduledExecutor();
+    private final long period;
+    private long lastDumpTimestamp;
+
+    @Inject
+    public JmxHistoryDumper(
+            JmxHistoricalData jmxHistoricalData,
+            JmxMetadata jmxMetadata,
+            JmxRecordSetProvider jmxRecordSetProvider,
+            JmxConnectorConfig jmxConfig)
+    {
+        this.jmxHistoricalData = requireNonNull(jmxHistoricalData, "jmxStatsHolder is null");
+        this.jmxMetadata = requireNonNull(jmxMetadata, "jmxMetadata is null");
+        this.jmxRecordSetProvider = requireNonNull(jmxRecordSetProvider, "jmxRecordSetProvider is null");
+        this.period = requireNonNull(jmxConfig, "jmxConfig is null").getDumpPeriod().toMillis();
+    }
+
+    public void start()
+    {
+        long nowMillis = getNowMillis();
+        long initialDelay = nowMillis - roundToPeriod(nowMillis + period);
+        executor.scheduleAtFixedRate(this, initialDelay, period, MILLISECONDS);
```

Comment:
`scheduleAtFixedRate` can cause multiple runs to happen back-to-back, after a long delay. E.g., if you set a period of 1s and the first run takes 5s, it will do 4 consecutive runs after the first one completes.

To illustrate:

```java
ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();

System.out.println("start: " + Instant.now());

AtomicBoolean flag = new AtomicBoolean();
executor.scheduleAtFixedRate(() -> {
    if (!flag.getAndSet(true)) {
        try {
            Thread.sleep(5000);
        }
        catch (InterruptedException e) {
            e.printStackTrace();
            Thread.currentThread().interrupt();
        }
    }
    System.out.println(Instant.now());
}, 0, 1, TimeUnit.SECONDS);
```

produces:

```
start: 2016-05-04T20:33:59.600Z
2016-05-04T20:34:04.663Z
2016-05-04T20:34:04.663Z
2016-05-04T20:34:04.663Z
2016-05-04T20:34:04.663Z
2016-05-04T20:34:04.663Z
2016-05-04T20:34:04.664Z
2016-05-04T20:34:05.661Z
2016-05-04T20:34:06.662Z
```